python/cuml/common/cuda.pyx

```diff
@@ -0,0 +1,74 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+
+class CudaRuntimeError(RuntimeError):
+    def __init__(self, extraMsg=None):
+        cdef _Error e = cudaGetLastError()
+        cdef bytes errMsg = cudaGetErrorString(e)
+        cdef bytes errName = cudaGetErrorName(e)
+        msg = "Error! %s reason='%s'" % (errName.decode(), errMsg.decode())
+        if extraMsg is not None:
+            msg += " extraMsg='%s'" % extraMsg
+        super(CudaRuntimeError, self).__init__(msg)
+
+
+cdef class Stream:
+    """
+    Stream represents a thin-wrapper around cudaStream_t and its operations.
+    """
+
+    # NOTE:
+    # If we store _Stream directly, this always leads to the following error:
+    #   "Cannot convert Python object to '_Stream'"
+    # I was unable to find a good solution to this in reasonable time. Also,
+    # since cudaStream_t is a pointer anyways, storing it as an integer should
+    # be just fine (although, that certainly is ugly and hacky!).
+    cdef size_t s
+
+    def __cinit__(self):
+        if self.s != 0:
+            return
+        cdef _Stream stream;
+        cdef _Error e = cudaStreamCreate(&stream)
+        if e != 0:
```

Comment:
I guess it is not possible to extract `CUDA_SUCCESS` from the CUDA C headers?

Reply:
I didn't come across a good way to wrap partial set of enum's for cython. Maybe @dantegd knows a way? If there's one, I agree with you. Creating such a thing would be cleaner.

Reply:
@dantegd @cjnolet do you guys know of a way to do this in cython?

Reply:
I believe Cython treats enums as integers, so you could likely get away with only defining the first few options in the enum if that's what you care about, and then you can compare against the enum instead of 0, but it's still somewhat brittle.

Reply:
Thanks @kkraus14. Yes, that's a bit brittle.

Reply:
@jirikraus this reminds me of the following "shortcut" I took to make sure we can read the `cudaError_t` return values from all cuda API calls. This can be found in `cuda.pxd` line:26. So, effectively, the enum returned by all such functions will be treated like an integer in the python world!

Reply:
Thanks @teju85 and @kkraus14 for the explanation. I am not really worried about the type conversion from enum to int. I pointed this out because I am worried in case `CUDA_SUCCESS` is ever changed to have another value than `0`. But I must admit that this is very very unlikely and I am probably overthinking this. This discussion can be resolved IMHO.

Reply:
Thanks @jirikraus . However, I still have a filed an issue #474 so that we don't forget this one.